Seek support for HLS and Progressive HTTP streams

LEChaney opened the pull request:
Implemented seek support for progressive HTTP stream protocol. This should work for any plugins using HTTPStream, HTTPSelect or SegmentedHTTPStream as their streaming method.
Also implemented multi-threaded streaming for progressive HTTP streams.

To enable seek support the player transport mode must be set to HTTP using either --player-http or --player-continuous-http.

Tested working with VLC.

LEChaney committed:
Added new stream type: SegmentedHTTPStream

Added alternative streaming protocol for handling progressive http streams.
The new protocol uses multiple segments to improve download speeds.
Added seghttp:// protocol option to cli to open streams using new protocol.

LEChaney committed:
Fixed a compatibility issue with older versions of python.

LEChaney committed:
Implemented Seek support for seghttp streams using player-continuous-http option
Broken: player-http option causes player/stream to immediately close

LEChaney committed:
Implemented response streaming to reduce buffer time
Broken: player-http option causes player/stream to immediately close

LEChaney committed:
Fixed a bug that could cause the program to block forever.

Fixed a bug that could cause the program to block forever if a connection error occurred while downloading a stream segment.

LEChaney committed:
Fixed a bug that could cause the program to crash

Fixed a bug that could cause the program to crash if a connection error occurred while downloading a stream segment.

LEChaney committed:
More error checking.

Should be a lot more tolerant of stream faults and timeouts now.

LEChaney committed:
Fix: Pending worker threads can no longer start work while the thread pool is closing.

LEChaney committed:
Download threads now handle close events that occur on separate threads

LEChaney committed:
Removed extraneous debug code

LEChaney committed:
Fixed some incorrect logging output

LEChaney committed:
Merge branch 'feature/stream-response' into bugfix/player-http

LEChaney committed:
Fixed broken --player-http option

Changed execution path to be similar to that of --player-continuous-http but with different flags set.
Will now exit only if the players http session is closed and not re-opened within a small time window.

LEChaney committed:
Fixed infinite retries when unable to open stream with --player-http option.

LEChaney committed:
Dummy Commit: Trigger travis build.

LEChaney committed:
Added seek support to standard HTTPStream:// protocol

LEChaney committed:
Removed seghttp://<url> cli protocol. HTTP stream type now automatically selected.

The httpstream://<url> protocol now selects the correct stream type automatically based on --stream-segment-threads option.
If this option is greater than 1 then theSegmentedHTTP stream type is used to parallelize the stream.

LEChaney committed:
Merge branch 'feature/httpseek' into develop

LEChaney committed:
Updated issues doc with requirements to enable HTTP seek.

LEChaney committed:
Merge branch 'feature/httpseek' into develop

LEChaney committed:
Dummy commit: Trigger travis rebuild.

LEChaney commented:
Future plans to make a plugin for AnimeLab which uses Progressive HTTP to stream video.
Also plan on implementing seek support for HLS streams, Crunchyroll in particular.

LEChaney committed:
Fixed all broken plugins that were receiving unexpected seek position argument (I hope.....)
Added HTTP stream type auto selection to plugins.

Plugins should now automatically select whether to use HTTPStream or SegmentedHTTPStream based on the value of the --stream-segment-threads option

LEChaney committed:
Revert of dirty fix for plugins that don't understand seek crashing on open.
Kept HTTP stream type auto selection change.

ian10951 commented:
Excellent work.

Seek when using Livestreamer will be a major enhancement.
However,when trying to stream I get this.

Got Exception:  **init**() got an unexpected keyword argument 'seek_pos'

This is using an enigma2 VU+ Solo2 and livestreamersrv

If I revert back to the development version without your latest commits the stream plays perfectly.

Ian.

LEChaney commented:
Thanks, is that the entire error you received? Our did you get a full stack traces well? I'm not sure where that init method is being called.

Unfortunately livestreamersrv won't currently work with seeking as it doesn't handle seek requests sent from the player. In order to make this work you currently have to get the first byte position from the http range header in the request sent by the player and pass it to the stream's open method. Since livestreamersrv is using its own http server it isn't passing this info onto the stream.

I currently have to pass the seek_pos parameter around to everything and make sure I don't pass it to anything that's not expecting it which I feel is very dirty. I'm working on proper event system to distribute seek events to subscribers 'only' which should fix issues like this, but if you have more details I can probably do a quick bug fix so at least it doesn't crash.

ian10951 commented:
It doesnt crash, it just doesnt stream.

The error was obtained from telnet using livestreamersrv manualstart
then selecting the stream on my VU+ Solo2.
Usually there's loads of info about the pipe ect but your version just gives

Got Exception: init() got an unexpected keyword argument 'seek_pos'

No other info at all.
It does this on both http streams and HLS streams. I didnt test HDS or rtmp

That's a real blow about seek not going to work when using the livestreamersrv deamon as its sadly been missing from enigma2 boxes when using Livestreamer and I do know that there are many,many enigma2 users who would like this feature.

Ah well never mind lol

Ian.

LEChaney committed:
Implemented Seek using new event system.

No longer shuts down and restarts all the streaming modules to perform a seek.

LEChaney committed:
Bugfix: Fixed an issue where download threads could modify each others http request headers.

A pass by reference was causing each download thread to have a reference to the same request header instead of each thread having its own. This meant it was sometimes possible for one thread to modify the request header of another before that thread was able to send its download request to the stream server. This could lead to garbage in the output buffer and result it stream glitches or crashes.

LEChaney committed:
Fixed some python 2.7 compatibility issues.

LEChaney committed:
Added seek event support for single threaded HTTP streams.

LEChaney committed:
Fixed a bug causing the old download thread to keep running after seek.

This affected single threaded HTTP streaming.

LEChaney committed:
Merge branch 'exp/seek-event' into feature/httpseek

LEChaney committed:
Merge branch 'feature/httpseek' into develop

# Conflicts:
#	.travis.yml

LEChaney committed:
Bugfix: Minor bugfix to get tests to pass

Each stream now has its own message broker instance. This allows multiple streams to be instantiated without encountering errors caused by trying to register mailboxes with the same handle on a global message broker.

LEChaney committed:
Merge branch 'feature/httpseek' into develop

ian10951 commented:
Just to let you know.

Your latest fixes now start and stream perfectly on an Enigma2 box using Livestreamersrv

Ian.

LEChaney committed:
Bugfix: Minor stability improvements

Filtering messages received by a mailbox by their source mailbox handle is now thread safe.

LEChaney committed:
Fixed some debug logging output

LEChaney committed:
Removed a seemingly unnecessary call to sleep in running property of PlayerOutput.

This call was responsible for 50% of the wait time of time to first byte after a seek event.

LEChaney committed:
Fixed a bug causing http server to receive incomplete requests.

LEChaney committed:
Merge branch 'feature/httpseek' into develop

LEChaney committed:
Fixed some compatibility issues with python 2.6

LEChaney committed:
Merge branch 'feature/httpseek' into develop

LEChaney committed:
Fixed some compatibility issues with python 3

LEChaney commented:
Thanks for the update. I might take a look at livestreamersrv at a later date once I have hls seek done. I'm a little confused at the moment as to why travis can't complete the python 2.6 build. All the tests run fine on my computer running a 2.6 environment....

ian10951 commented:
I'm surprised anyone still uses Python 2.6 lol

That would be awesome if you could do that and bring the Livestreamersrv up to date as its not been modified in 3 years. Loads of people would be very grateful indeed.

Ian.

LEChaney committed:
Merge branch 'feature/httpseek' into develop

LEChaney committed:
Added instance of the MessageBroker to stream baseclass instead of just HTTP stream types.

LEChaney committed:
Merge branch 'feature/httpseek' into develop

LEChaney committed:
Stopping the player now stops the stream.
Various fixes so shutdown or stop behaves appropriately.

LEChaney committed:
Merge branch 'feature/httpseek' into develop

LEChaney committed:
Initial seek support for HLS streams.
Need to fix stream shutdown and improve performance.

LEChaney committed:
Updated streaming response object to be more generic.

LEChaney committed:
HLS segments now stream instead of downloading the entire segment.

HLS segments now stream out to the player instead of waiting for the entire segment to download before outputting, this results in a significant reduction in stream delay and seek time but can lead to buffering when first opening the stream or performing a seek.
Added timestamps to debug level logging output.

LEChaney committed:
End of stream events are now handled properly
Accelerated initial fetch of meta data used for seek.

LEChaney committed:
Added new stream end handling to single threaded http streams.

LEChaney committed:
Changed the logging test to reflect new timestamped format.

LEChaney commented:
Finished implementing HLS seek support. Also accelerated stream loading significantly. Streams should now open much faster and have less stream delay on live streams.

ian10951 commented:
Nice One.

HLS Streams now pause, FF and RW using an enigma2 box and the Livestreamersrv

However,like you said earlier,for seek to work modifications will also be required to that file too,so at present thats still not working.

Live Streams do seem to significantly start faster too.

Sterling work !

LEChaney committed:
Fixed a dependency issue in Python 2.6

ian10951 commented:
There does however seem to be a problem with http streams which do play perfectly when using the chrippa Livestreamer version. This is the same on other http streams too,and not just with the example ive copied and pasted here.

Heres the log,but unfortunately it doesnt explain much other than showing that the stream doesnt play and eventually times out.

EDIT:

Hmmm

When I paste the log this site refuses to display 

SegmentedHTTPStream('http://92.222.205.40:8085/lighd') before the URL

Which is what the log in my telnet window actually shows

OpenPLi 4 vusolo2

vusolo2 login: root
root@vusolo2:~# livestreamersrv stop
root@vusolo2:~# livestreamersrv manualstart
Mon May 16 09:47:23 2016 Server Starts - :88
localhost.localdomain - - [16/May/2016 09:47:32] "GET /httpstream://http://92.22
2.205.40:8085/lighd HTTP/1.1" 200 -
localhost.localdomain - - [16/May/2016 09:47:32] URL: httpstream://http://92.222
.205.40:8085/lighd Quality: best
Streams: ['live', 'worst', 'best']
<SegmentedHTTPStream('http://92.222.205.40:8085/lighd')>
localhost.localdomain - - [16/May/2016 09:48:02] "GET /httpstream://http://92.22
2.205.40:8085/lighd HTTP/1.1" 200 -
localhost.localdomain - - [16/May/2016 09:48:02] URL: httpstream://http://92.222
.205.40:8085/lighd Quality: best
Streams: ['live', 'worst', 'best']
<SegmentedHTTPStream('http://92.222.205.40:8085/lighd')>
localhost.localdomain - - [16/May/2016 09:48:32] "GET /httpstream://http://92.22
2.205.40:8085/lighd HTTP/1.1" 200 -
localhost.localdomain - - [16/May/2016 09:48:32] URL: httpstream://http://92.222
.205.40:8085/lighd Quality: best
Streams: ['live', 'worst', 'best']
<SegmentedHTTPStream('http://92.222.205.40:8085/lighd')>

## Got Exception:  Read timeout

Exception happened during processing of request from ('127.0.0.1', 51775)
Traceback (most recent call last):
  File "/usr/lib/python2.7/SocketServer.py", line 582, in process_request_thread
  File "/usr/lib/python2.7/SocketServer.py", line 323, in finish_request
  File "/usr/lib/python2.7/SocketServer.py", line 640, in **init**
  File "/usr/lib/python2.7/SocketServer.py", line 693, in finish
  File "/usr/lib/python2.7/socket.py", line 303, in flush

## error: [Errno 32] Broken pipe

LEChaney committed:
Added seek support for EVENT type vods (ie Twitch VODs)

LEChaney committed:
Closing player now shuts down livestreamer cleanly when using standard input pipe transport mode.
Livestreamer now shuts down faster when the player is closed using HTTP transport mode.

LEChaney committed:
Updated the issues doc to reflect  HLS seek support.

LEChaney commented:
Ah, it's because I check for content length in the HTTP headers.

Here are the headers from that server:
HTTP/1.0 200 OK
Content-Length: 0
Content-type: application/octet-stream
Cache-Control: no-cache

It's also missing the accepts ranges header, so seek and multi-segmented downloading is impossible. I'll try to make it fall back to single threaded downloading if the content-length is zero or the accepts-range header is missing.

In the mean time you can solve this by setting the stream-segment-threads option to 1 so that the single threaded streaming method is used.

ian10951 commented:
Yes thats cured it for now.

Thank You.

LEChaney committed:
Falls back to single segment downloading when stream cannot be segmented.

LEChaney committed:
Minor stability fix, tests now pass.

LEChaney committed:
Fixed a bug that was causing deadlock when the ringbuffer was full.

L2501 commented:
could be useful to add automatic http-reconnect to live streams for players that dont support it
vlc has --http-reconnect but it does not seem to cache properly

LEChaney commented:
Does --player-continuous-http not work for you? It should keep trying to reconnect if the stream goes down.

L2501 commented:
it does "work" but there is a delay between [cli][info] Stream ended and [cli][info] Opening stream due to the vlc cache maybe? 

`livestreamer -p 'vlc --http-reconnect' --ringbuffer-size 50M --stream-segment-threads 2 --player-continuous-http 'httpstream://http://149.202.195.182:2500/live/basit/basit/2969.ts' best`

LEChaney commented:
I think I see what you mean. To be honest, I'm not really working on this anymore as it was taking up too much of my time. There are definitely still some bugs, but they happen rarely (only when I'm trying to break it so far). The issue is probably related to player read and segment download maximum retry attempts and timeouts. I've done some initial testing and I seem to get deadlock when I try to simulate the stream going offline with --http-reconnect. Are you getting any timeouts/retries in the console?

L2501 commented:
livestreamer -l debug -p 'vlc --http-reconnect' --ringbuffer-size 50M --stream-segment-threads 2 --player-continuous-http 'httpstream://http://cp.dmbshare.net:8000/live/master99/master99/4594.ts' best
[2016-07-19 12:37:31.492177][cli][info] Found matching plugin stream for URL httpstream://http://cp.dmbshare.net:8000/live/master99/master99/4594.ts
[2016-07-19 12:37:31.494504][stream.http][debug] Retrieving complete content length
[2016-07-19 12:37:31.624741][stream.http][debug] Unable to get content length
[2016-07-19 12:37:31.626259][cli][info] Available streams: live (best, worst)
[2016-07-19 12:37:31.627131][cli][info] Starting player: vlc --http-reconnect --file-caching=1
[2016-07-19 12:37:31.934168][cli][info] Got HTTP request from VLC/2.2.4 LibVLC/2.2.4
[2016-07-19 12:37:32.034616][cli][info] Opening stream: live (http)
[2016-07-19 12:37:32.035931][stream.http][debug] Retrieving complete content length
[2016-07-19 12:37:32.264528][stream.http][debug] Unable to get content length
[2016-07-19 12:37:34.761133][cli][debug] Pre-buffering 8192 bytes
[2016-07-19 12:37:34.761841][cli][debug] Writing stream to player
[2016-07-19 12:37:44.896137][cli][info] Stream ended
[2016-07-19 12:38:00.979270][cli][info] Got HTTP request from VLC/2.2.4 LibVLC/2.2.4
[2016-07-19 12:38:01.079891][stream.http][debug] Retrieving complete content length
[2016-07-19 12:38:01.443432][stream.http][debug] Unable to get content length
[2016-07-19 12:38:01.446709][cli][info] Opening stream: live (http)
[2016-07-19 12:38:01.447545][stream.http][debug] Retrieving complete content length
[2016-07-19 12:38:01.563176][stream.http][debug] Unable to get content length
[2016-07-19 12:38:02.061424][cli][debug] Pre-buffering 8192 bytes
[2016-07-19 12:38:02.062252][cli][debug] Writing stream to player
[2016-07-19 12:38:12.240100][cli][info] Stream ended
[2016-07-19 12:38:22.476753][cli][info] Got HTTP request from VLC/2.2.4 LibVLC/2.2.4
[2016-07-19 12:38:22.578235][stream.http][debug] Retrieving complete content length
[2016-07-19 12:38:22.698486][stream.http][debug] Unable to get content length
[2016-07-19 12:38:22.698797][cli][info] Opening stream: live (http)
[2016-07-19 12:38:22.699390][stream.http][debug] Retrieving complete content length
[2016-07-19 12:38:22.809072][stream.http][debug] Unable to get content length
[2016-07-19 12:38:23.308229][cli][debug] Pre-buffering 8192 bytes
[2016-07-19 12:38:23.309717][cli][debug] Writing stream to player
[2016-07-19 12:38:31.849659][cli][info] Stream ended

looks like it is not going to work nicely unless it does some internal stream muxing. there is a kodi script that tries to do something like that https://github.com/Shani-08/ShaniXBMCWork2/blob/master/script.video.F4mProxy/lib/TSDownloader.py works almost decently all things considered.

LEChaney commented:
Can you confirm whether things are working as expected without my changes? I'm not really sure what's happening with that stream, it goes through a bunch of 302 redirects then ends on a 404 for me.

Also segmented streams won't work unless there is a known content length (for HTTP progressive streams) as discussed above. It's different for HLS streams because there are multiple segments available with known sizes even for live streams, so each of them can downloaded in parallel. With an HTTP progressive stream I have know way of knowing how far ahead the video has been buffered server side, or whether the server will even respond appropriately to a byte range request unless I see the accepts ranges header in the server response.

L2501 commented:
sry it's unrelated to your changes, thats just how those specific "demo" streams work, they disconnect you after few secs but allow unlimited connects so might be possible to leech it live, but seems it's not really viable to make it work smoothly without missing keyframes.

LEChaney commented:
Some issues with my current code and possible improvements:
- Errors when seeking all the way to the end of a VOD stream.
- Can take a long time to open HLS streams with many segments, as I need the complete size information, which can only be acquired by prefetching the headers for every segment in the stream. This can take a long time for streams with small segments (like twitch VODs). This is probably very hard to fix without a big architectural change to the way livestreamer core returns stream data (make it a proxy server for different stream types, players then interpret the stream type (i.e. HLS) properly and can make proper seek requests without needing the content length of the entire VOD).
- Currently creates an additional layer of buffering, one for each _queued_ segment (the queue size is 20). All of these buffers can download and fill, regardless of whether the main buffer is full or not. The segment threads option doesn't affect this (once the thread is finished it is free to download the next segment in the queue, which has it's own buffer to fill). This means the real buffer size is something like the main (ringbuffer) buffer size + 20 \* segment-buffer-size. It would be better (albeit a lot more complex) to have one segmented buffer that everything can read into in a parallel manner without overwriting each other's data, and which blocks when full.
- Currently all buffered data is thrown out once a seek request is received, even if it could be reused. The two situations where we should be able to reuse data are: When we seek ahead, but not past the point where we don't have any data. When we seek backwards within some buffered window in which data is saved for seek purposes.

Possible architectural improvements:
- Change livestreamer core to act as a proxy server, with only HTTP output. This way other applications which implement their own servers currently, no longer have to, and don't have to worry about forwarding content size information in order for seek support to work.
- Mailbox communication system should be changed to work like Golang channels for consistency/simplicity. 
- A lot of code could be cleaned up by moving more into the stream type base classes.
- Some of the HLS segmented stream stuff should be rewritten so I'm not hacking mutable fields onto an immutable type to store stream length data.
